sssd_test_framework/utils/adcli.py

```diff
@@ -0,0 +1,198 @@
+"""Perform actions on Active Directory."""
+
+from __future__ import annotations
+
+from pytest_mh import MultihostHost, MultihostUtility
+from pytest_mh.cli import CLIBuilder, CLIBuilderArgs
+from pytest_mh.conn import ProcessResult
+from pytest_mh.utils.fs import LinuxFileSystem
+
+__all__ = [
+    "ADCLIUtils",
+]
+
+
+class ADCLI(MultihostUtility[MultihostHost]):
+    """
+    Call commands from adcli
+    """
+
+    def __init__(self, host: MultihostHost, fs: LinuxFileSystem) -> None:
+        super().__init__(host)
+
+        self.cli: CLIBuilder = self.host.cli
+        """Command line builder."""
+
+        self.fs: LinuxFileSystem = fs
+        """Filesystem utils."""
+
+    def _info(
+        self,
+        *,
+        domain: str | None = None,
+        domain_controller: str | None,
+    ) -> str:
+        """
+        Call ``adcli info `` with given arguments.
+
+        :param domain: Displays discovered information about an Active Directory domain, defaults to None
+        :type domain: str | None,
+        :param domain_controller: Domain controller to connect, defaults to None
+        :type domain_controller: str | None, optional
+        """
+        args: CLIBuilderArgs = {
+            "domain": (self.cli.option.VALUE, domain),
+            "domain_controller": (self.cli.option.VALUE, domain_controller),
+        }
+
+        self.host.conn.exec(["adcli", "info"] + self.cli.args(args))
+
+    def _testjoin(
+        self,
+        *,
+        domain: str| None = None,
+        domain_controller: str| None = None,
+        host_keytab: str| None = None,
+    ) -> str:
+        """
+        call ``adcli testjoin `` with given arguments.
+
+        :param domain: Target Active Directory domain, defaults to None
+        :type domain: str | None, optional
+        :param domain_controller: Domain controller to connect
+        :type domain_controller: str | None, optional
+        """
+        args: CLIBuilderArgs = {
+            "domain": (self.cli.option.VALUE, domain),
+            "domain_controller": (self.cli.option.VALUE, domain_controller),
+            "host_keytab": (self.cli.option.VALUE, host_keytab),
+        }
+        self.host.conn.exec(["adcli", "testjoin"] + self.cli.args(args))
+
+    def _join(
+        self,
+        *,
+        domain: str| None = None,
```

Comment:
There should be probably either domain or domain_controller.  
The manual page states that it expects domain so I would go with that.
Getting both None or both populated is a recipe for a disaster.

Reply:
rearranged.